tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+    TZRMJD             56330.1763862586450231
+    TZRSITE                                 8
+    TZRFRQ                           1547.743
+    """
+    model = get_model(StringIO(par_base))
+    return model
+
+
+@pytest.fixture()
+def make_toas_to_go_with_two_piece_model(build_piecewise_model_with_two_pieces):
+    # makes toas to go with the two non-overlapping, complete coverage model
+    m_piecewise = deepcopy(build_piecewise_model_with_two_pieces)
+    m_piecewise.setup()
+    toas = pint.simulation.make_fake_toas_uniform(
+        m_piecewise.START.value + 1, m_piecewise.FINISH.value - 1, 20, m_piecewise
+    )  # slightly within group edges to make toas unambiguously contained within groups
+    return toas
+
+
+def get_toa_group_indexes(model, toas):
+    # returns array of group indexes associated with each toa. (i.e. which group is each toa in)
+    model.setup()
+    index = model.which_group_is_toa_in(toas)
+    return index
+
+
+def get_number_of_groups(model):
+    # returns number of groups
+    model.setup()
+    number_of_groups = model.get_number_of_groups()
+    print(model.as_parfile())
+    print(model.get_number_of_groups())
+    return number_of_groups
+
+
+@pytest.fixture()
+def build_piecewise_model_with_two_pieces(build_model):
+    # takes the basic model frame and adds 2 non-ovelerlapping pieces to it
+    piecewise_model = deepcopy(build_model)
+    piecewise_model.remove_range(0)
+    piecewise_model.setup()
+    spacing = (piecewise_model.FINISH.value - piecewise_model.START.value) / 2
+    for i in range(0, 2):
+        piecewise_model.add_group_range(
+            piecewise_model.START.value + spacing * i,
+            piecewise_model.START.value + spacing * (i + 1),
+            j=i,
+        )
+        piecewise_model.add_piecewise_param(
+            "A1", "ls", piecewise_model.A1.value + i * 1e-5, i
+        )
+        piecewise_model.add_piecewise_param(
+            "T0", "d", piecewise_model.T0.value + i * 1e-5, i
+        )
+    return piecewise_model
+
+
+def test_add_piecewise_parameter(build_model):
+    # test: see if the model can be reproduced after piecewise parameters have been added,
+    # checks by comparing parameter keys in both the old and new file. Should have the number of matches = number of parameters
+    m_piecewise = deepcopy(build_model)
+    n = 10
+    m_piecewise.remove_range(0)
+    for i in range(0, n):
+        m_piecewise.add_group_range(
+            m_piecewise.START.value + 501 * i, m_piecewise.START.value + 1000 * i, j=i
+        )
+        m_piecewise.add_piecewise_param("A1", "ls", m_piecewise.A1.value + 1, i)
+        m_piecewise.add_piecewise_param("T0", "d", m_piecewise.T0.value, i)
+    m3 = get_model(StringIO(m_piecewise.as_parfile()))
+    param_dict = m_piecewise.get_params_dict(which="all")
+    copy_param_dict = m3.get_params_dict(which="all")
+    number_of_keys = 0
+    comparison = 0
+    for key, value in param_dict.items():
```

Comment:
Can you just ask `assert param_dict == copy_param_dict`? If this fails you'll get a list of the mismatched keys/values.